crates/namada/src/vm/wasm/memory.rs

```diff
@@ -220,62 +240,69 @@ fn check_bounds(memory: &Memory, base_addr: u64, offset: usize) -> Result<()> {
 
 /// Read bytes from memory at the given offset and length
 fn read_memory_bytes(
+    store: &mut impl wasmer::AsStoreMut,
     memory: &Memory,
     offset: u64,
     len: usize,
 ) -> Result<Vec<u8>> {
-    check_bounds(memory, offset, len)?;
-    let offset = usize::try_from(offset)?;
-    let vec: Vec<_> = memory.view()[offset..checked!(offset + len)?]
-        .iter()
-        .map(|cell| cell.get())
-        .collect();
-    Ok(vec)
+    check_bounds(store, memory, offset, len)?;
+    let mut buf = vec![0; len];
+    memory.view(&store.as_store_mut()).read(offset, &mut buf)?;
+    Ok(buf)
 }
 
 /// Write bytes into memory at the given offset
 fn write_memory_bytes(
+    store: &mut impl wasmer::AsStoreMut,
     memory: &Memory,
     offset: u64,
     bytes: impl AsRef<[u8]>,
 ) -> Result<()> {
-    let slice = bytes.as_ref();
-    let len = slice.len();
-    check_bounds(memory, offset, len as _)?;
-    let offset = usize::try_from(offset)?;
-    memory.view()[offset..checked!(offset + len)?]
-        .iter()
-        .zip(slice.iter())
-        .for_each(|(cell, v)| cell.set(*v));
+    let buf = bytes.as_ref();
+    check_bounds(store, memory, offset, buf.len() as _)?;
+    memory.view(&store.as_store_mut()).write(offset, buf)?;
     Ok(())
 }
 
 /// The wasm memory
-#[derive(Debug, Clone, Default)]
+#[derive(Debug, Clone)]
 pub struct WasmMemory {
-    pub(crate) inner: LazyInit<wasmer::Memory>,
+    store: Rc<RefCell<Store>>,
+    memory: Rc<RefCell<Option<wasmer::Memory>>>,
 }
 
+// TODO: Wasm memory is neither `Send` nor `Sync`, but we must implement
```

Comment:
should we remove the `TODO: ` prefix and just leave it as a note?

Reply:
ah I forgot to update this `TODO`. #3313 covers this